packages/devtools/lib/src/timeline/timeline_protocol.dart

```diff
@@ -322,22 +397,120 @@ class TimelineEvent {
+        findChild(e, childName);
+      }
+    }
+
+    findChild(this, childName);
+    return childFound;
+  }
+
+  /// Adds a child event.
+  ///
+  /// If we know the child's location (knownLocation == true)
+  /// add the child directly via [_addChild]. If we do not know the child's
+  /// location, find it's proper place in the subtree and add the child there.
+  void addChild(TimelineEvent child, {bool findChildLocation = false}) {
+    bool newChildAdded = false;
+
+    // Places the child in it's correct position amongst the other children.
+    void _putChildInSubtree(TimelineEvent root, TimelineEvent newChild) {
+      // [root] is a leaf. Add child here.
+      if (root.children.isEmpty) {
+        root._addChild(child);
+        newChildAdded = true;
+      }
+
+      if (newChildAdded) return;
+
+      final List<TimelineEvent> _children =
+          List<TimelineEvent>.from(root.children);
+      for (int i = 0; i < _children.length; i++) {
+        final TimelineEvent currentChild = _children[i];
+
+        // [newChild] is the parent of [currentChild].
+        if (newChild._isParentOf(currentChild)) {
+          // Link [currentChild] with its correct parent [newChild].
+          newChild._addChild(currentChild);
+
+          // Unlink [child] from its incorrect parent [this].
+          root.children.remove(currentChild);
+
+          // Link [newChild] with its correct parent [this].
+          if (!newChildAdded) _addChild(newChild);
+
+          newChildAdded = true;
+        }
+
+        // [child] is the parent of [newChild].
+        if (currentChild._isParentOf(newChild)) {
+          // Recurse on [currentChild]'s subtree.
+          _putChildInSubtree(currentChild, newChild);
+        }
+      }
+
+      // If we have not added the child at this point, [currentChild] and
+      // [newChild] are siblings.
+      if (!newChildAdded) {
+        root._addChild(newChild);
+        newChildAdded = true;
+        return;
+      }
+    }
+
+    if (findChildLocation) {
+      _putChildInSubtree(this, child);
+    } else {
+      _addChild(child);
+    }
+  }
+
+  void _addChild(TimelineEvent child) {
+    children.add(child);
+    child.parent = this;
+  }
+
+  // TODO(kenzie): consider comparing with an epsilon for endTime.
+  bool _isParentOf(TimelineEvent e) {
+    if (endTime != null && e.endTime != null) {
+      return startTime < e.startTime && endTime > e.endTime;
+    } else {
+      return startTime < e.startTime;
+    }
+  }
+
   void format(StringBuffer buf, String indent) {
-    buf.writeln('$indent$name [${startTime}u]');
+    buf.writeln(
+        '$indent$name [start: $startTime] [end: $endTime] [dur: $duration]');
     for (TimelineEvent child in children) {
       child.format(buf, '  $indent');
     }
   }
 
   @override
-  String toString() => '$name, start=$startTime duration=$duration';
+  String toString() => '[$type] $name [start $startTime] [end $endTime] [dur '
```

Comment:
bonus points: use DiagnosticableTreeMixin to provide better debugging output for this
it would also let you write golden tests verifying tree output a bit more easily. 
That class is already imported into scope. Let me know if you have questions.

Reply:
Added a TODO for a follow up CL